### ISSUE TITLE
XProc harness does not work (or incomplete documentation)

### ISSUE THREAD

**Issue by AirQuick** [xspec/xspec#74]
I just tried the XProc harness for XSpec, and I think the code is somewhat broken or at least the documentation is incomplete. Here's what happened to me:

As I know little about XProc, I started from the documentation. Per [Running Scenarios](https://github.com/xspec/xspec/wiki/Running-Scenarios), this is the basic command line for XProc:

> java com.xmlcalabash.drivers.Main -i source=my-tests.xspec saxon-xslt-harness.xproc

So, on the latest master + XML Calabash 1.1.15-97, I tried this, but it failed:

```winbatch
P:\xspec-master>java -cp xmlcalabash-1.1.15-97.jar com.xmlcalabash.drivers.Main -i source=tutorial/escape-for-regex.xspec src/harnesses/saxon/saxon-xslt-harness.xproc
ERROR: err:XC0011:Could not load http://www.jenitennison.com/xslt/xspec/generate-xspec-tests.xsl (file:/P:/xspec-master/src/harnesses/harness-lib.xpl) dtd-validate=false
ERROR: err:XC0011:Could not load http://www.jenitennison.com/xslt/xspec/generate-xspec-tests.xsl (file:/P:/xspec-master/src/harnesses/harness-lib.xpl) dtd-validate=false
ERROR: Unknown error
```

Looks like `$xspec-home` is missing, so I added the parameter to the command line. It made some progress, but still didn't complete:

```winbatch
P:\xspec-master>java -cp xmlcalabash-1.1.15-97.jar com.xmlcalabash.drivers.Main -i source=tutorial/escape-for-regex.xspec -p xspec-home=file:///P:/xspec-master/ src/harnesses/saxon/saxon-xslt-harness.xproc
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:Testing with SAXON HE 9.7.0.11
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:No escaping
...
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:Strings should be escaped and status attributes should be added
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:      FAILED
ERROR: src/harnesses/harness-lib.xpl:238:35:Input source unbound on p:error step named !1.5.7.2.1 and no default binding available.
ERROR: err:XS0032:Input source unbound on p:error step named !1.5.7.2.1 and no default binding available.
ERROR: It is a static error if no connection is provided and the default readable port is undefined.
```

Out of a mere guess, I modified `src/harnesses/harness-lib.xpl:238` as follows:

```xml
        <p:error code="t:ERR001">
          <p:input port="source">
            <p:inline>
              <message>t:report not found!</message>
            </p:inline>
          </p:input>
        </p:error>
```

It made some more progress, but still failed:

```winbatch
P:\xspec-master>java -cp xmlcalabash-1.1.15-97.jar com.xmlcalabash.drivers.Main -i source=tutorial/escape-for-regex.xspec -p xspec-home=file:///P:/xspec-master/ src/harnesses/saxon/saxon-xslt-harness.xproc
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:Testing with SAXON HE 9.7.0.11
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:No escaping
...
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:Strings should be escaped and status attributes should be added
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:      FAILED
ERROR: err:XC0011:Could not load http://www.jenitennison.com/xslt/xspec/format-xspec-report.xsl (file:/P:/xspec-master/src/harnesses/harness-lib.xpl) dtd-validate=false
ERROR: err:XC0011:Could not load http://www.jenitennison.com/xslt/xspec/format-xspec-report.xsl (file:/P:/xspec-master/src/harnesses/harness-lib.xpl) dtd-validate=false
ERROR: Unknown error
```

Looks like `$xspec-home` is missing from `<p:declare-step type="t:format-report" name="format">`. So I modified it so that it resembles `<p:declare-step type="t:compile-xslt" name="compile-xsl">` with regard to `xspec-home`.
Finally it worked:

```winbatch
P:\xspec-master>java -cp xmlcalabash-1.1.15-97.jar com.xmlcalabash.drivers.Main -i source=tutorial/escape-for-regex.xspec -p xspec-home=file:///P:/xspec-master/ src/harnesses/saxon/saxon-xslt-harness.xproc
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:Testing with SAXON HE 9.7.0.11
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:No escaping
...
INFO : src/harnesses/saxon/saxon-xslt-harness.xproc:22:45:      FAILED
INFO : src/harnesses/harness-lib.xpl:230:38:passed: 5 / pending: 0 / failed: 1 / total: 6
<html xmlns:test="http://www.jenitennison.com/xslt/unit-test"
      xmlns="http://www.w3.org/1999/xhtml">
   <head>
      <title>Test Report for P:/xspec-master/tutorial/escape-for-regex.xslt (5/0/1/6)</title>
...
      </div>
   </body>
</html>
```

**Comment by cirulls** [comment]
Thanks for highlighting this, I came to the same conclusion at about the same time when checking #72 raised by @opax. 

As I wrote in #72, I suggest to use the implementation in Oxygen. Would you agree with that?

**Comment by cirulls** [comment]
I started working on this in branch 72, see in particular [.travis.yml](https://github.com/xspec/xspec/blob/72/.travis.yml) and [saxon-xslt-harness.xproc](https://github.com/xspec/xspec/blob/72/src/harnesses/saxon/saxon-xslt-harness.xproc). The Travis configuration is currently running a single test just to check that the XProc harness executes correctly but it would be good to let it run all the tests executed by the test suite. I'll keep working on it in the next days. 

I assign this issue to me but feel free to give your feedback so far.

**Comment by AirQuick** [comment]
So, are you reverting the whole `src/harnesses` back to [v0.3](https://github.com/xspec/xspec/tree/77213c795263f88c6e46ed52ac33a42f057878d0/src/harnesses)?
Or only the Saxon harness?

**Comment by cirulls** [comment]
Good point! I feel like it is an issue similar to the one we encountered in #26 when a regression bug from v0.4.0-RC was identified. 

As far as I can see, the XProc harness for Saxon does not work as it is in the code base. The one implemented in Oxygen based on v0.3.0 does work and @opax introduced an improvement for the encoding in #72. My priority would be to make the Saxon XProc harness work and document how to run it. I have not tested the other harnesses and, apart from eXist-db, I am not very familiar with the other tools (MarkLogic, Zorba, BaseX). However, I suspect that these harnesses may be problematic as well as they all rely on variable ```$xspec-home``` being passed as you identified it for the Saxon harness. In any case, there is no documentation on how to run them so this should definitely be improved. 

@AirQuick : what is your take on this?

**Comment by AirQuick** [comment]
As you pointed out, this look similar to #26. Apparently `src/harnesses` was refactored and released as RC in haste.

The commit logs for `src/harnesses` suggest that after v0.3.0, the biggest changes were the refactoring for multiple processors (3cd1a4e81ad5eb60e9407d8bc3a5b8517eaf3272) and the BaseX server support (9f600d58600941c5824a1795fb5ba63d26060d40). Other few changes look like minor tweaks.

The refactoring is the issue in question here. The BaseX server support does not seem to be tested very well anyway... So I think we should reset `src/harness` to v0.3.0 and let the code base stay there (and take minor fixes like #72), until someone gets interested seriously in improving the multiple processors support.

Regarding documentation and priority, I totally agree with you.

**Comment by cirulls** [comment]
Thanks for your feedback. I'll do the following:

- Revert ```src/harnesses``` to v0.3.0
- Implement the pull request submitted by @opax in #72 
- Write a test that checks the encoding
- Add the Saxon XProc harness to the ```.travis``` configuration so that it is included in our test suite (I'll see if I can make it run all the tests)
- Document how to run the Saxon XProc harness in the wiki

When I'm done I'll submit a pull request and request a review (@opax for the encoding part, @AirQuick for the Travis part). 

@AirQuick : sorry if I haven't yet reviewed your pull requests or responded to other tickets, I read your comments but I need a bit more time to properly assess them and say something sensible about it!

**Comment by cirulls** [comment]
I guess they started integrating v0.5.0 in their beta release. Few weeks ago I emailed @alexJitianu mentioning that there were few bugs we noticed and fixed after the XSpec v0.5.0 release, I hope this message will prompt him again (alternatively I can drop him another email).

**Comment by AlexJitianu** [comment]
Hi,
I guess I can make a build from the current development stream. We usually integrate releases since they are usually tested and more reliable... I'm not sure if it will make it in v19 but if it doesn't then I will schedule it for a maintenance build.
